File: day03.zig
const std = @import("std");
const print = std.debug.print;
const input = @embedFile("inputs/day03.txt");

pub fn main() !void {
    var lines: usize = 0;
    var totals: [12]usize = undefined;
    std.mem.set(usize, totals[0..], 0);

    var iterate = std.mem.tokenize(input, "\n");
    while (iterate.next()) |line| {
        for (line) |c, i| {
            totals[i] += c-'0';
        }
        lines += 1;
    }

    var gamma: i32 = 0;
    var epsilon: i32 = 0;
    var power: i32 = 1;
    for (totals) |_, i| {
        const v = totals[(totals.len-1)-i];
        if (v > lines/2) {
            gamma += power;
        } else {
            epsilon += power;
        }
        power *= 2;
    }
    print("part1: {d}", .{gamma*epsilon});
}

File: day03.py
with open('inputs/day03.txt') as f:
    data = f.read().splitlines()

def get_common_bits(data):
    bits = [list() for _ in range(len(data[0]))]
    for line in data:
        for i, char in enumerate(line):
            bits[i].append(char)
    return ''.join([max(set(li), key=li.count) for li in bits])

gamma = get_common_bits(data)
epsilon = ''.join('1' if x == '0' else '0' for x in gamma)
print('part 1:', int(gamma,2)*int(epsilon,2))

def part2(flip=False):
    d = data[:]
    x = 0
    while len(d) != 1:
        cb = get_common_bits(d)
        if flip:
            remove = [line for line in d if line[x] == cb[x]]
        else:
            remove = [line for line in d if line[x] != cb[x]]
        for r in remove:
            d.remove(r)
        x += 1
    return int(d[0], 2)

print('part 2:', part2()*part2(flip=True))
